### PR TITLE
[10.0][ADD] analytic tag dimension warning for purchase and sale

### PR DESCRIPTION
**Pull request by grindtildeath**
Purchase Analytic Tag Dimension
===========================

This module extends the functionality of analytic_tag_dimension to have the
warning appear while saving purchase orders, should two tags from the same
dimension be used.

Sale Analytic Tag Dimension
=======================

This module extends the functionality of analytic_tag_dimension to have the
warning appear while saving sale orders, should two tags from the same
dimension be used.

### CODE REVIEW

**Review by jcoux — APPROVED** [review]
Code seems good (not tested) :+1:

**Review by rami-wafaie — APPROVED** [review]
Tested it, looks fine to me
:+1:

File: sale_analytic_tag_dimension/README.rst
Comment (leemannd) [inline sale_analytic_tag_dimension/README.rst]: '..., should two tags from the same dimension be used' A bit unclear to me

**Review by leemannd — APPROVED** [review]
Thanks. LGTM!

**Comment by grindtildeath** [comment]
@pedrobaeza Commits squashed, ready to merge :)

File: purchase_analytic_tag_dimension/README.rst
Comment (LoisRForgeFlow) [inline purchase_analytic_tag_dimension/README.rst]: s/should/whenever ?

File: purchase_analytic_tag_dimension/README.rst
Comment (LoisRForgeFlow) [inline purchase_analytic_tag_dimension/README.rst]: Please remove this section or at least the template figure.

File: purchase_analytic_tag_dimension/__init__.py
Comment (LoisRForgeFlow) [inline purchase_analytic_tag_dimension/__init__.py]: add ending blankline and header with at least the coding `# -*- coding: utf-8 -*-`

File: purchase_analytic_tag_dimension/models/__init__.py
Comment (LoisRForgeFlow) [inline purchase_analytic_tag_dimension/models/__init__.py]: Same as above, and for all init files.

File: purchase_analytic_tag_dimension/__manifest__.py
Comment (lreficent) [inline purchase_analytic_tag_dimension/__manifest__.py]: Space before Odoo

File: purchase_analytic_tag_dimension/__manifest__.py
Comment (LoisRForgeFlow) [inline purchase_analytic_tag_dimension/__manifest__.py]: remove empty tags

File: sale_analytic_tag_dimension/README.rst
Comment (LoisRForgeFlow) [inline sale_analytic_tag_dimension/README.rst]: Same comments as in above readme

File: sale_analytic_tag_dimension/__manifest__.py
Comment (LoisRForgeFlow) [inline sale_analytic_tag_dimension/__manifest__.py]: remove

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]
Code review

File: purchase_analytic_tag_dimension/__init__.py
Comment (yvaucher) [reply purchase_analytic_tag_dimension/__init__.py]: Here there is no code, common sense is that we don't need copyright reminder in this file.

**Comment by grindtildeath** [comment]
Hello @lreficent 

Wanted to write this on the mailing list, but I think it's more appropriate here.
It's nothing against you, but I think this (as well as https://github.com/OCA/product-attribute/pull/274#pullrequestreview-68517866) is typically the kind of reviews which are not needed, and restrain people trying to contribute to the OCA. 

No hard feelings, but IMO your expertise might be better used on other PRs... especially when these were approved at least two times.

File: purchase_analytic_tag_dimension/__init__.py
Comment (LoisRForgeFlow) [reply purchase_analytic_tag_dimension/__init__.py]: I'm not talking about the copyrights, I refer to coding and even license as [OCA template](https://github.com/OCA/maintainer-tools/blob/master/template/module/models/__init__.py) suggest.

**Comment by LoisRForgeFlow** [comment]
Hi @grindtildeath 

I think the module is ok functionally since, as you said, there are already other approvals. Also note, that it is a small and not very complex module so I reckon that, since you have decided to share it with OCA, you should take the ocassion to try to follow as much as possible the OCA guidelines and make it perfect. This way, when you contribute next time you will be more used to the conventions and your code will be better, and next time again and so on.

I know it can be bothering to change this little details, but I also know that It doesn't take more than 2 minutes. After all, my next actions would probably be to approve your code and your odds to get it merge will be increased. Isn't it a good deal?

And finally to not extend too much, regarding the other review you pointed out, **I gave my approval** but just told you as well some little change you could made, is it so hurting?

**Review by LoisRForgeFlow — APPROVED** [review]
BTW, I almost forgot to approve, I was never my goal to block it

**Comment by grindtildeath** [comment]
@lreficent 
No worries and thanks for your review anyway. 
It's not hurting at all, but sometimes it gets frustrating to have reviews such as 'add a space between OCA authors' on a module with almost no code, when there are a few more PRs opened in this same repository which don't get any.
Also I'm sorry if my comment was a bit harsh, as it wasn't meant to be :wink:

**Comment by rafaelbn** [comment]
Hi @grindtildeath ,

Thank you very much for this contribution! 😄 

Please take my comment as an effort to understand how to help OCA and improve situation (in general) for getting more contributors (functional and technical) to the OCA. Please don't miss understand me and imagine you are the reviewer and @lreficent is the contributor 😆 . I please because I don't want misunderstandings and I'm not blocking this PR which **only miss a review from a PSC of this project to be merged!** ☮️ 

Also take in care that I agree with you in

> sometimes it gets frustrating to have reviews such as 'add a space between OCA authors' on a module with almost no code, when there are a few more PRs opened in this same repository which don't get any

He have discussed in the OCA Code Sprint about get more contributor: both functional and technical. @elicoidal and @StefanRijnhart lead this initiative. I'm sorry using this PR as an example of, but it's a good one. We/You/Everyone would like to have more contributors! (more **reviewers** in fact is needed).

- I think [OCA Guidelines](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md) are easy to understand and quickly to read.
- There is a [section in the OCA Guidelines](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#it-makes-sense-to-be-picky-in-the-following-cases) that say 

> "It makes sense to be picky in the following cases"

 with a point 

> "Respect of Odoo/community conventions"

This is very important for the OCA IMHO.

Why I say this?

Because you said:

> the kind of reviews which are not needed, and restrain people trying to contribute to the OCA 

And I think comment like this also retrain people trying to contribute to the OCA **reviewing**. We should try to understand both sides.

In the section [There are the following important parts in a review](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#there-are-the-following-important-parts-in-a-review) we must add **Thanks the contributor / reviewer for their work. Someone has take time to review your work and try to give you some tips to improve the work, so be thankful for that.**

I agree with @jgrandguillaume about Open Doors in OCA, we were discussing about this topic also. We have to open for both, developers and funcional consultants (which make your contributions are used by more people!)

# About this PR

I used to think that it must be done a single PR per module. What do you think? Imagine you review a PR with two modules in a single PR. Are you going to please ask for make a single PR per module?  You should as the review could be a Business Analyst without any look in the code.

I try to tested it but runbot was not available so now it is rebuilding

Thanks you for this great modules! 👍 

cc @angelmoya @carlosdauden @sergio-teruel maybe you would like to take a look in this 2 modules.

Really guys now I'm afraid about making this comment! 😱 I'm your friend don't hate me! Not sure if press the button or delete everthing and go to sleep....

File: purchase_analytic_tag_dimension/__manifest__.py
Comment (rafaelbn) [inline purchase_analytic_tag_dimension/__manifest__.py]: I miss the category like in https://github.com/OCA/account-analytic/blob/10.0/analytic_multicurrency/__manifest__.py#L26

File: sale_analytic_tag_dimension/__manifest__.py
Comment (rafaelbn) [inline sale_analytic_tag_dimension/__manifest__.py]: Also in this module I miss the category as in https://github.com/OCA/account-analytic/blob/10.0/analytic_multicurrency/__manifest__.py#L26

**Review by rafaelbn — COMMENTED** [review]
I think the name of the modules should be:

- analytic_tag_dimension_purchase
- analytic_tag_dimension_sale

Futhermore, as the purpose ot both modules is:

> You can not set two tags from same dimension.

I recomend a name for both modules more specific like

- analytic_tag_dimension_purchase_unique_tag
- analytic_tag_dimension_sale_unique_tag

or similar.

**Comment by rafaelbn** [comment]
Functionally tested 👍 

But I insist that names are wrong. Maybe...

- analytic_tag_dimension_purchase_warning
- analytic_tag_dimension_sale_warning

This modules just warn before what the main module has forbidden https://github.com/OCA/account-analytic/blame/10.0/analytic_tag_dimension/README.rst#L17

**Comment by grindtildeath** [comment]
@rafaelbn 
No worries and thanks for your review. I just changed the module names as I agree with your last suggestion.
BTW I did add both modules in the same PR, because they are basically the same, and I thought it would be easier to have all the reviews at the same place, even if it might be circumventing the convention a little bit.

As I wrote, I'm sorry if my first comment was a bit harsh, being rude to @lreficent never was my intent, and I ought to reread myself before commenting.  :fearful:

**Review by rafaelbn — APPROVED** [review]

**Pull request merged**
